devices/usbc/ctrl.c

```diff
@@ -0,0 +1,410 @@
+	if (outQh) {
+		dtd = ctrl_allocQtdMem();
+		if (dtd == NULL)
+			return ERR_MEM;
+
+		ctrl_common.dc->endptqh[qh].base = (((u32)dtd) & ~0xfff);
+		ctrl_common.dc->endptqh[qh].size = 0x40;
+		ctrl_common.dc->endptqh[qh].head = dtd;
+		ctrl_common.dc->endptqh[qh].tail = dtd;
+	}
+
+	if (inQH) {
+		dtd = ctrl_allocQtdMem();
+		if (dtd == NULL)
+			return ERR_MEM;
+
+		ctrl_common.dc->endptqh[++qh].base = (((u32)dtd) & ~0xfff) + (64 * sizeof(dtd_t));
+		ctrl_common.dc->endptqh[qh].size = 0x40;
+		ctrl_common.dc->endptqh[qh].head = dtd + 64;
+		ctrl_common.dc->endptqh[qh].tail = dtd + 64;
+	}
+
+	return ERR_NONE;
+}
+
+
+static int ctrl_initEndptQh(int endpt, int dir, endpt_data_t *endpt_init)
+{
+	u32 setup = 0;
+	int qh = endpt * 2 + dir;
+
+	ctrl_allocBuff(endpt, dir);
+
+	ctrl_common.dc->endptqh[qh].caps =  endpt_init->caps[dir].max_pkt_len << 16;
+	ctrl_common.dc->endptqh[qh].caps |= endpt_init->caps[dir].ios << 15;
+	ctrl_common.dc->endptqh[qh].caps |= endpt_init->caps[dir].zlt << 29;
+	ctrl_common.dc->endptqh[qh].caps |= endpt_init->caps[dir].mult << 30;
+	ctrl_common.dc->endptqh[qh].dtd_next = 1;
+
+	setup |= endpt_init->ctrl[dir].data_toggle << (6 + dir * 16);
+	setup |= endpt_init->ctrl[dir].type << (2 + dir * 16);
+
+	return setup;
+}
+
+
+int ctrl_endptInit(int endpt, endpt_data_t *endpt_init)
+{
+	s8 i;
+	int res = ERR_NONE;
+	u32 setup = 0;
+
+	if (endpt == 0)
+		return -1;
+
+	if ((res = ctrl_dtdInit(endpt, endpt_init->caps[USB_ENDPT_DIR_IN].init, endpt_init->caps[USB_ENDPT_DIR_OUT].init)) != ERR_NONE)
+		return res;
+
+	for (i = 0; i < ENDPOINTS_DIR_NB; ++i) {
+		if (endpt_init->caps[i].init)
+			setup |= ctrl_initEndptQh(endpt, i, endpt_init);
+	}
+
+	*(ctrl_common.dc->base + endptctrl0 + endpt) = setup;
+
+	for (i = 0; i < ENDPOINTS_DIR_NB; ++i) {
+		if (endpt_init->caps[i].init)
+			*(ctrl_common.dc->base + endptctrl0 + endpt) |= 1 << (7 + i * 16);
+	}
+
+	return res;
+}
+
+
+int ctrl_endpt0Init(void)
+{
+	int i;
+	u32 qh_addr;
+
+	if (ctrl_allocBuff(0, USB_ENDPT_DIR_IN) < 0)
+		return ERR_MEM;
+
+	if (ctrl_allocBuff(0, USB_ENDPT_DIR_OUT) < 0)
+		return ERR_MEM;
+
+	ctrl_common.data->endpts[0].caps[USB_ENDPT_DIR_IN].init = 1;
+	ctrl_common.data->endpts[0].caps[USB_ENDPT_DIR_OUT].init = 1;
+
+	/* map queue head list */
+	ctrl_common.dc->endptqh = usbclient_allocBuff(USB_BUFFER_SIZE);
+
+	if (ctrl_common.dc->endptqh == NULL)
+		return ERR_MEM;
+
+	for (i = 0; i < USB_BUFFER_SIZE; ++i)
+		((char *)ctrl_common.dc->endptqh)[i] = 0;
+
+	qh_addr = ((u32)((void *)ctrl_common.dc->endptqh)) & ~0xfff;
+
+	ctrl_common.dc->endptqh[0].caps =  0x40 << 16; /* max 64 bytes */
+	ctrl_common.dc->endptqh[0].caps |= 0x1 << 29;
+	ctrl_common.dc->endptqh[0].caps |=  0x1 << 15; /* ios */
+	ctrl_common.dc->endptqh[0].dtd_next = 0x1; /* invalid */
+
+	ctrl_common.dc->endptqh[1].caps =  0x40 << 16;
+	ctrl_common.dc->endptqh[1].caps |= 0x1 << 29;
+	ctrl_common.dc->endptqh[1].caps |=  0x1 << 15;
+	ctrl_common.dc->endptqh[1].dtd_next = 1;
```

Comment:
```suggestion
	u32 tmp;
	tmp =  0x40 << 16 | 0x1 << 29 | 0x1 << 15;
	ctrl_common.dc->endptqh[0].caps =  tmp;
	ctrl_common.dc->endptqh[0].dtd_next = 0x1; /* invalid */

	ctrl_common.dc->endptqh[1].caps =  tmp;
	ctrl_common.dc->endptqh[1].dtd_next = 1;
```
And the second proposal. Arguments above.
For order either `0x1` or '1'.

Reply:
Done